knowledge-base/adr/0038-csrf-middleware.md

```diff
@@ -0,0 +1,181 @@
+## Status
+
+accepted
+
+### Structure of the Code Before Cahnge
+
+The CSRF token handling is done in the `http-client` before the request is executed:
+
+```ts
+    //http-client.ts
+  request.headers = await addCsrfTokenToHeader(destination, request, options);
+
+    return executeWithMiddleware(
+      requestConfig.middleware,
+      {
+        ...
+      },
+      () => executeFn(request)
+    );
+```
+
+Currently, the middleware is only added to the `executeFn()` to the target system.
+For the fetch request it is not considered.
+Note that any kind of resilience entering via the middleware is not considered.
+
+### General Improvement I - Make Arguments Adjustable
+
+The CSRF token revealed a problem.
+Currently, the middleware contains the function `fn` with no way to change the parameters:
+
+```ts
+{
+    fn: () => Promise<ReturnType>,
+    context: Context
+}
+```
+
+With the CSRF token you would like to change the arguments of the axios request (add the token) without rebuilding the full `fn`.
+Note that the `fn` may contain already middleware layers from before.
+Same is true for the expiring access tokens from a destination cache.
+
+An idea could be, to add the argument type and a reference to the `fn`:
+
+```ts
+interface Function<ReturnType,ArguemntType> {
+
+    (): Promise<ReturnType>,
+    arg: ArguemntType
```

Comment:
I am wondering why this has to be a separate property. Could it be in the fn signature?

`(arg: ArgumentType): Promise<ReturnType>` ?

Reply:
We make it a argument of the function and add a reference to the context so that it can be changed.